### PR TITLE
Make sure drawImage width and height are not too big

### LINKED ISSUE #4337: svg icon in image style not drawn in Safari if offset is used
This is similar to: http://stackoverflow.com/questions/29633355/canvas-context-drawimage-differences-across-browsers

Changing the draw instruction to:

```
         context.drawImage(image, originX, originY, width - originX, height - originY,
                x, y, width * pixelRatio, height * pixelRatio);
```

fixes this issue and the icon is drawn correctly.

### PR DESCRIPTION
**Pull request by bartvde**
in the case of an offset.

fixes #4337

### CODE REVIEW

**Comment by ahocevar** [comment]
It would be nice to have a render test for this.

**Comment by bartvde** [comment]
But the problem is only visible in Safari? So how would a render test help?

**Comment by ahocevar** [comment]
Does this fix break things on other browsers?

In theory, if you create the render test reference image in a different browser, and run the test in Safari, it should fail. In practice, the resemble.js image comparison is often too picky. Maybe the render tests should use Pixelmatch instead. See #4288.

**Comment by bartvde** [comment]
No it's still fine on other browsers (I checked Chrome and FF).

There is still no way to run the render tests on OSX right?

**Comment by ahocevar** [comment]
It would at least be nice to have a passing render test before this change, to make sure that it doesn't break things on other browsers.

**Comment by bartvde** [comment]
Agreed, okay I'll add one.

**Comment by ahocevar** [comment]
> There is still no way to run the render tests on OSX right?

They run fine in the browser (http://localhost:3000/test_rendering/. In El Capitan with latest XCode, they also run with `make test-rendering` now.

**Comment by bartvde** [comment]
Ah great, that's good to know.

**Comment by bartvde** [comment]
@ahocevar and they currently pass all for you on OSX? I get 5 failures:

```
Barts-MacBook-Air:ol3_bartvde bartvandeneijnden$ make test-rendering
node tasks/test-rendering.js
Gecko: 33.1.1
ol.rendering.layer.VectorTile vector tile layer renders correctly with the canvas renderer
Error: Error: expected '2.12' to be below 1.5 (http://127.0.0.1:3001/node_modules/expect.js/index.js:96)

ol.rendering.reproj.Tile tile projection from EPSG:4326 to EPSG:3857
Error: Error: expected '7.42' to be below 1.5 (http://127.0.0.1:3001/node_modules/expect.js/index.js:96)

ol.rendering.reproj.Tile dateline wrapping wraps X when prime meridian is shifted
Error: Error: expected '6.20' to be below 1.5 (http://127.0.0.1:3001/node_modules/expect.js/index.js:96)

ol.rendering.reproj.Tile dateline wrapping displays north pole correctly (EPSG:3413)
Error: Error: expected '2.69' to be below 1.5 (http://127.0.0.1:3001/node_modules/expect.js/index.js:96)

ol.rendering.style.RegularShape #render tests the canvas renderer
Error: Error: expected '6.36' to be below 6 (http://127.0.0.1:3001/node_modules/expect.js/index.js:96)

5 test(s) failed.
```

**Comment by bartvde** [comment]
Added a rendering test for an icon with an offset, let's see if Travis also thinks it's okay.

**Comment by bartvde** [comment]
Hmm, but of course my change does influence the resulting image. So a rendering image created before this change is different.

**Comment by bartvde** [comment]
So this needs more work.

**Comment by bartvde** [comment]
Okay have a fix now that results in the same reference image before and after this change.

**Comment by bartvde** [comment]
@ahocevar this is now really ready for review.

File: src/ol/render/canvas/canvasreplay.js [hunk 1]
```diff
@@ -343,8 +343,9 @@ ol.render.canvas.Replay.prototype.replay_ = function(
             context.globalAlpha = alpha * opacity;
           }
 
-          context.drawImage(image, originX, originY, width, height,
-              x, y, width * pixelRatio, height * pixelRatio);
+          context.drawImage(image, originX, originY, width - originX,
+              height - originY, x, y, (width - originX) * pixelRatio,
+              (height - originX) * pixelRatio);
```
Comment (marcjansen) [inline src/ol/render/canvas/canvasreplay.js]: Very very minor: For readability it'd be nice if thes eypressions would be variables. No must, of course

**Comment by marcjansen** [comment]
I added in comment which you may want to take. Otherwise this looks good to me. I'd favor if @ahocevar looked at this as well.

Thanks, Bart.

File: src/ol/render/canvas/canvasreplay.js (hunk 1 above)
Comment (bartvde) [inline src/ol/render/canvas/canvasreplay.js]: Yes makes sense, will do.

**Comment by ahocevar** [comment]
Looking at this now.

**Comment by ahocevar** [comment]
Looks great now, thanks @bartvde. Please merge.

**Pull request merged**